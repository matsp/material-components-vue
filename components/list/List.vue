<template>
  <ul
    v-if="tag === 'ul'"
    :class="classes"
    class="mdc-list"
    v-bind="$attrs"
    @MDCList:action="onAction"
    @_init="onParentInit"
  >
    <slot />
  </ul>
  <nav
    v-else-if="tag === 'nav'"
    :class="classes"
    class="mdc-list"
    v-bind="$attrs"
    @MDCList:action="onAction"
    @_init="onParentInit"
  >
    <slot />
  </nav>
  <div
    v-else-if="tag === 'div'"
    :class="classes"
    class="mdc-list"
    v-bind="$attrs"
    @MDCList:action="onAction"
    @_init="onParentInit"
  >
    <slot />
  </div>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '../base'
import { MDCList } from '@material/list'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  inject: {
    getList: {
      default: null
    }
  },
  model: {
    prop: 'selectedIndex',
    event: 'change'
  },
  props: {
    avatar: {
      type: Boolean,
      default: false
    },
    dense: {
      type: Boolean,
      default: false
    },
    twoLine: {
      type: Boolean,
      default: false
    },
    nonInteractive: {
      type: Boolean,
      default: false
    },
    singleSelection: {
      type: Boolean,
      default: false
    },
    vertical: {
      type: Boolean,
      default: true
    },
    wrapFocus: {
      type: Boolean,
      default: true
    },
    selectedIndex: {
      type: [Number, Array],
      default: -1
    },
    tag: {
      type: String,
      default: 'ul'
    },
    js: {
      type: Boolean,
      default: true
    }
  },
  data () {
    return {
      mdcList: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-list--dense': this.dense,
        'mdc-list--two-line': this.twoLine,
        'mdc-list--avatar-list': this.avatar,
        'mdc-list--non-interactive': this.nonInteractive
      }
    }
  },
  watch: {
    singleSelection () {
      if (this.singleSelection) {
        this.$el.setAttribute('role', 'listbox')
        this.$slots.default.forEach((n) => {
          if (n.elm instanceof HTMLElement) {
            n.elm.setAttribute('role', 'option')
          }
        })
      }
      if (this.mdcList) { this.mdcList.singleSelection = this.singleSelection }
    },
    vertical () {
      if (this.mdcList) { this.mdcList.vertical = this.vertical }
    },
    wrapFocus () {
      if (this.mdcList) { this.mdcList.wrapFocus = this.wrapFocus }
    },
    selectedIndex () {
      if (this.mdcList && ((this.selectedIndex instanceof Number && this.selectedIndex > -1) || this.selectedIndex instanceof Array)) this.mdcList.selectedIndex = this.selectedIndex
    },
    js () {
      if (!(this.getList instanceof Function)) {
        this.reInstantiateList()
      } else {
        console.warn('Prop \'js\' is omitted when your <m-list> is placed inside a menu or drawer. The \'js\' prop won\'t take effect.')
      }
    },
    tag () {
      this.$nextTick(() => {
        if (!(this.getList instanceof Function)) {
          this.reInstantiateList()
        } else {
          console.warn('Prop \'tag\' is not reactive when your <m-list> is placed inside a menu or drawer. Changing \'tag\' won\'t take effect.')
        }
      })
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    if (!(this.getList instanceof Function)) { // can not be init by parent
      this.instantiate()
    }
  },
  beforeDestroy () {
    if (this.mdcList) this.mdcList.destroy()
  },
  methods: {
    instantiate () {
      this.mdcList = MDCList.attachTo(this.$el)
      this.propsSetting()
    },
    propsSetting () {
      this.mdcList.singleSelection = this.singleSelection
      if (this.singleSelection) this.$el.setAttribute('role', 'listbox')
      this.mdcList.vertical = this.vertical
      if ((this.selectedIndex instanceof Number && this.selectedIndex > -1) || this.selectedIndex instanceof Array) this.mdcList.selectedIndex = this.selectedIndex
      this.mdcList.wrapFocus = this.wrapFocus
    },
    /**
     * only called if there's no injected getList method
     */
    reInstantiateList () {
      if (this.mdcList instanceof MDCList) {
        this.mdcList.destroy()
      }
      this.mdcList = undefined
      if (this.js) { // prop js can be safely checked
        this.instantiate()
      }
    },
    updateSlot () {
      if (this.mdcList instanceof MDCList) this.mdcList.layout()
      if (this.singleSelection) {
        this.$slots.default.forEach((n) => {
          if (n.elm instanceof HTMLElement) {
            n.elm.setAttribute('role', 'option')
          }
        })
      }
    },
    onAction (e) {
      if (this.mdcList instanceof MDCList && this.mdcList.selectedIndex >= 0) {
        this.$emit('change', this.mdcList.selectedIndex)
      }
      this.$emit('action', e.detail)
    },
    onParentInit () {
      const list = this.getList()
      if (list instanceof MDCList) {
        if (this.mdcList instanceof MDCList) this.mdcList.destroy()
        this.mdcList = list
        this.propsSetting()
      }
    },
  }
}
</script>
