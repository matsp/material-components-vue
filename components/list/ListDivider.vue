<template>
  <li
    :class="classes"
    role="separator"
    class="mdc-list-divider"
  />
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    inset: {
      type: Boolean,
      default: false
    },
    padded: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes () {
      return {
        'mdc-list-divider--inset': this.inset,
        'mdc-list-divider--padded': this.padded
      }
    }
  }
}
</script>
