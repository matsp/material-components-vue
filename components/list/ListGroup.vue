<template>
  <div class="mdc-list-group">
    <h3
      v-if="subheader.length > 0"
      class="mdc-list-group__subheader"
    >
      {{ subheader }}
    </h3>
    <slot />
  </div>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    subheader: {
      type: String,
      default: ''
    }
  }
}
</script>
