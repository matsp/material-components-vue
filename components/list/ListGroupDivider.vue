<template>
  <hr class="mdc-list-divider">
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin]
}
</script>
