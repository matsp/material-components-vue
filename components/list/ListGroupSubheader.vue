<script>
import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    level: {
      type: Number,
      default: 3,
      validator: function (value) {
        return value > 0 && value < 7
      }
    }
  },
  render: function (createElement) {
    return createElement(
      'h' + this.level, // tag name
      {
        class: {
          'mdc-list-group__subheader': true
        }
      }, this.$slots.default // array of children
    )
  }
}
</script>
