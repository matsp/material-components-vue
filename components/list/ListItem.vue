<template>
  <a
    v-if="href"
    :aria-selected="activated"
    :class="classes"
    :href="href"
    class="mdc-list-item"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <slot
      v-if="$slots['graphic']"
      name="graphic"
    />
    <slot />
    <span
      v-if="$slots['text'] || $slots['primaryText'] || $slots['secondaryText']"
      class="mdc-list-item__text"
    >
      <slot name="text" />
      <span
        v-if="$slots['primaryText']"
        class="mdc-list-item__primary-text"
      >
        <slot name="primaryText" />
      </span>
      <span
        v-if="$slots['secondaryText']"
        class="mdc-list-item__secondary-text"
      >
        <slot name="secondaryText" />
      </span>
    </span>
    <slot
      v-if="$slots['meta']"
      name="meta"
    />
  </a>
  <li
    v-else
    :class="classes"
    class="mdc-list-item"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <slot name="graphic" />
    <slot />
    <span
      v-if="$slots['text'] || $slots['primaryText'] || $slots['secondaryText']"
      class="mdc-list-item__text"
    >
      <slot name="text" />
      <span
        v-if="$slots['primaryText']"
        class="mdc-list-item__primary-text"
      >
        <slot name="primaryText" />
      </span>
      <span
        v-if="$slots['secondaryText']"
        class="mdc-list-item__secondary-text"
      >
        <slot name="secondaryText" />
      </span>
    </span>
    <slot name="meta" />
  </li>
</template>

<script>
import { MDCRipple } from '@material/ripple'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    activated: {
      type: Boolean,
      default: false
    },
    selected: {
      type: Boolean,
      default: false
    },
    href: {
      type: String,
      default: ''
    },
    disabled: {
      type: Boolean,
      default: false
    },
    ripple: {
      type: Boolean,
      default: true
    }
  },
  data () {
    return {
      mdcRipple: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-list-item--activated': this.activated,
        'mdc-list-item--selected': this.selected,
        'mdc-list-item--disabled': this.disabled
      }
    }
  },
  watch: {
    classes () {
      this.reInstantiateRipple()
      if (this.selected) this.$el.setAttribute('aria-selected', 'true')
    },
    ripple () {
      this.reInstantiateRipple()
    }
  },
  mounted () {
    this.updateSlots()
    this.slotObserver = new MutationObserver(() => this.updateSlots())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    if (this.ripple) this.mdcRipple = MDCRipple.attachTo(this.$el)
    if (this.selected) this.$el.setAttribute('aria-selected', 'true')
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
  },
  methods: {
    updateSlots () {
      if (this.$slots.graphic) {
        this.$slots.graphic.map(n => {
          n.elm.classList.add('mdc-list-item__graphic')
        })
      }
      if (this.$slots.meta) {
        this.$slots.meta.map(n => {
          n.elm.classList.add('mdc-list-item__meta')
        })
      }
    },
    reInstantiateRipple () {
      if (this.ripple) {
        if (this.mdcRipple) {
          this.mdcRipple.destroy()
        }
        MDCRipple.attachTo(this.$el)
      } else {
        if (this.mdcRipple) {
          this.mdcRipple.destroy()
        }
        this.mdcRipple = undefined
      }
    }
  }
}
</script>
