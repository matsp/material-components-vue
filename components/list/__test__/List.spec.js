import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import List from '../List'
import ListDivider from '../ListDivider'
import ListGroup from '../ListGroup'
import ListGroupDivider from '../ListGroupDivider'
import ListGroupSubheader from '../ListGroupSubheader'
import ListItem from '../ListItem'

describe('List', () => {
  it('should mount', () => {
    const wrapper = mount(List)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(List)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list')
  })

  it('should render as dense', () => {
    const wrapper = mount(List, {
      propsData: {
        dense: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list--dense')
  })

  it('should render as two line', () => {
    const wrapper = mount(List, {
      propsData: {
        twoLine: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list--two-line')
  })

  it('should render as avatar', () => {
    const wrapper = mount(List, {
      propsData: {
        avatar: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list--avatar-list')
  })

  it('should render as non interactive', () => {
    const wrapper = mount(List, {
      propsData: {
        nonInteractive: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list--non-interactive')
  })

  it('should render as nav', () => {
    const wrapper = mount(List, {
      propsData: {
        tag: 'nav'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('nav').classes()).toContain('mdc-list')
  })

  it('should render as div', () => {
    const wrapper = mount(List, {
      propsData: {
        tag: 'div'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('div').classes()).toContain('mdc-list')
  })
})

describe('List Divider', () => {
  it('should mount', () => {
    const wrapper = mount(ListDivider)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ListDivider)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-divider')
  })

  it('should render as inset', () => {
    const wrapper = mount(ListDivider, {
      propsData: {
        inset: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-divider--inset')
  })

  it('should render as padded', () => {
    const wrapper = mount(ListDivider, {
      propsData: {
        padded: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-divider--padded')
  })
})

describe('List Group', () => {
  it('should mount', () => {
    const wrapper = mount(ListGroup)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ListGroup)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-group')
  })
})

describe('List Group Divider', () => {
  it('should mount', () => {
    const wrapper = mount(ListGroupDivider)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ListGroupDivider)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-divider')
  })

  it('should render as inset', () => {
    const wrapper = mount(ListGroupDivider, {
      propsData: {
        inset: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-divider--inset')
  })

  it('should render as padded', () => {
    const wrapper = mount(ListGroupDivider, {
      propsData: {
        padded: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-divider--padded')
  })
})

describe('List Group Subheader', () => {
  it('should mount', () => {
    const wrapper = mount(ListGroupSubheader)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ListGroupSubheader)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-group__subheader')
  })

  it('should render as h1', () => {
    const wrapper = mount(ListGroupSubheader, {
      propsData: {
        level: 1
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('h1').classes()).toContain('mdc-list-group__subheader')
  })
})

describe('List Item', () => {
  it('should mount', () => {
    const wrapper = mount(ListItem)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ListItem)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-item')
  })

  it('should render as activated', () => {
    const wrapper = mount(ListItem, {
      propsData: {
        activated: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-item--activated')
  })

  it('should render as selected', () => {
    const wrapper = mount(ListItem, {
      propsData: {
        selected: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-item--selected')
  })

  it('should render as disabled', () => {
    const wrapper = mount(ListItem, {
      propsData: {
        disabled: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-list-item--disabled')
  })
})
