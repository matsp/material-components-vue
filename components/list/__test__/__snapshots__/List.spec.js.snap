// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Divider should render as inset 1`] = `<li role="separator" class="mdc-list-divider mdc-list-divider--inset"></li>`;

exports[`List Divider should render as padded 1`] = `<li role="separator" class="mdc-list-divider mdc-list-divider--padded"></li>`;

exports[`List Divider should render with no prop 1`] = `<li role="separator" class="mdc-list-divider"></li>`;

exports[`List Group Divider should render as inset 1`] = `<hr class="mdc-list-divider mdc-list-divider--inset">`;

exports[`List Group Divider should render as padded 1`] = `<hr class="mdc-list-divider mdc-list-divider--padded">`;

exports[`List Group Divider should render with no prop 1`] = `<hr class="mdc-list-divider">`;

exports[`List Group Subheader should render as h1 1`] = `<h1 class="mdc-list-group__subheader"></h1>`;

exports[`List Group Subheader should render with no prop 1`] = `<h3 class="mdc-list-group__subheader"></h3>`;

exports[`List Group should render with no prop 1`] = `
<div class="mdc-list-group">
  <!---->
</div>
`;

exports[`List Item should render as activated 1`] = `
<li class="mdc-list-item mdc-list-item--activated">
  <!---->
</li>
`;

exports[`List Item should render as disabled 1`] = `
<li class="mdc-list-item mdc-list-item--disabled">
  <!---->
</li>
`;

exports[`List Item should render as selected 1`] = `
<li class="mdc-list-item mdc-list-item--selected" aria-selected="true">
  <!---->
</li>
`;

exports[`List Item should render with no prop 1`] = `
<li class="mdc-list-item">
  <!---->
</li>
`;

exports[`List should render as avatar 1`] = `<ul class="mdc-list mdc-list--avatar-list"></ul>`;

exports[`List should render as dense 1`] = `<ul class="mdc-list mdc-list--dense"></ul>`;

exports[`List should render as div 1`] = `<div class="mdc-list"></div>`;

exports[`List should render as nav 1`] = `<nav class="mdc-list"></nav>`;

exports[`List should render as non interactive 1`] = `<ul class="mdc-list mdc-list--non-interactive"></ul>`;

exports[`List should render as two line 1`] = `<ul class="mdc-list mdc-list--two-line"></ul>`;

exports[`List should render with no prop 1`] = `<ul class="mdc-list"></ul>`;
